server/gui/workers.py

```diff
@@ -1,47 +1,94 @@
+from multiprocessing import Process
 import traceback
+import time
 
-from server.gui.utils import WorkerSignals
+import requests
 
+from server.gui.utils import SiteReadySignals
 
-class DataLoadWorker():
-    def __init__(self, data_file, layout="umap", *args, **kwargs):
-        super(DataLoadWorker, self).__init__()
+
+class EmittingProcess(Process):
+    def __init__(self, parent_conn, child_conn, *arg, **kwargs):
+        super(EmittingProcess, self).__init__()
+        self.parent_conn = parent_conn
+        self.child_conn = child_conn
+
+    def run(self):
+        self.parent_conn.close()
+
+    def emit(self, signal_name, *args):
+        message = (signal_name, *args)
+        self.child_conn.send(message)
+
+
+class Worker(EmittingProcess):
+    def __init__(self, parent_conn, child_conn, data_file, title, host, port, layout=["umap"], *args, **kwargs):
+        super(Worker, self).__init__(parent_conn, child_conn)
         self.data_file = data_file
         self.layout = layout
-        self.signals = WorkerSignals()
+        self.title = title
+        self.host = host
+        self.port = port
 
     def run(self):
+        super(Worker, self).run()
         if not self.data_file:
-            self.signals.finished.emit()
+            self.emit("finished")
             return
-
-        # delayed import to speed load
+        from server.app.app import Server
         from server.app.scanpy_engine.scanpy_engine import ScanpyEngine
-        args = {
-            "layout": self.layout,
-            "diffexp": "ttest",
-            "max_category_items": 100,
-            "diffexp_lfc_cutoff": 0.01,
-            "obs_names": None,
-            "var_names": None,
-        }
+        # create server
         try:
-            data_results = ScanpyEngine(self.data_file, args)
+            server = Server()
+            server.create_app()
         except Exception as e:
             traceback.print_exc()
-            self.signals.error.emit(str(e))
-        else:
-            self.signals.result.emit(data_results)
+            self.emit("server_error", str(e))
+            self.emit("finished")
+            return
+        # load data
+        try:
+            args = {
+                "layout": self.layout,
+                "diffexp": "ttest",
```

Comment:
`diffexp` config is obsolete/unused (AFAIK)

Reply:
I delete, but I've got updating diffexp in the gui in a future PR.